src/Interpreters/Cache/QueryConditionCache.cpp

```diff
@@ -16,51 +16,52 @@ QueryConditionCache::QueryConditionCache(const String & cache_policy, size_t max
 {
 }
 
-std::optional<QueryConditionCache::MatchingMarks> QueryConditionCache::read(const UUID & table_id, const String & part_name, size_t condition_hash)
+void QueryConditionCache::write(
+    const UUID & table_id, const String & part_name, size_t predicate_hash,
+    const MarkRanges & mark_ranges, size_t marks_count, bool has_final_mark)
 {
-    Key key = {table_id, part_name, condition_hash};
+    Key key = {table_id, part_name, predicate_hash};
 
-    if (auto entry = cache.get(key))
-    {
-        ProfileEvents::increment(ProfileEvents::QueryConditionCacheHits);
+    auto entry = std::make_shared<Entry>(marks_count);
 
-        std::lock_guard lock(entry->mutex);
-        return {entry->matching_marks};
-    }
+    /// The input mark ranges are the areas which the scan can skip later on.
+    for (const auto & mark_range : mark_ranges)
+        std::fill(entry->matching_marks.begin() + mark_range.begin, entry->matching_marks.begin() + mark_range.end, false);
+
+    if (has_final_mark)
+        entry->matching_marks[marks_count - 1] = false;
 
-    ProfileEvents::increment(ProfileEvents::QueryConditionCacheMisses);
+    LOG_DEBUG(
+        logger,
+        "Wrote entry for table_id: {}, part_name: {}, predicate_hash: {}, marks_count: {}, has_final_mark: {}, (ranges: {})",
+        table_id,
+        part_name,
+        predicate_hash,
+        marks_count,
+        has_final_mark,
+        toString(mark_ranges));
 
-    return std::nullopt;
+    auto load_func = [&](){ return entry; };
+    auto result = cache.getOrSet(key, load_func);
```

Comment:
What if it already exists in cache? Can we override some information? I guess it can't happen because for same condition and same part, same mark ranges will always be set, but I would like to confirm it.
Why don't we use simple `cache.set` instead?

Reply:
Nevermind ... I misunderstood the code. I reverted the commit which removes the locking and added a comment for the next person.